Implement RichIterable.groupByAndCollect().

Refer to discussion on [pull request 37](https://github.com/eclipse/eclipse-collections/pull/37) for more information. Also, please use the most up-to-date name decided.

I am trying to implement `groupByAndCollect` method on `RichIterable`. My test is failing with errors that looks confusing.
Here are some errors:
1. [ERROR]   `ImmutableSortedBagImplTest.RichIterable_groupByAndCollect` expected: `org.eclipse.collections.impl.map.mutable.UnifiedMap<{false=[12, 10, 8, 6, 4], true=[11, 9, 7, 5, 3]}>` but was: `org.eclipse.collections.impl.map.mutable.UnifiedMap<{false=[12, 10, 8, 6, 4], true=[11, 9, 7, 5, 3]}>`
2. [ERROR]   `ImmutableHashBiMapInverseTest.RichIterable_groupByAndCollect` expected:`<{false=[8, 10, 12, 4, 6], true=[9, 11, 3, 5, 7]}>` but was:`<{false=[4, 6, 8, 10, 12], true=[3, 5, 7, 9, 11]}>`

In case 1, both expected result and the actual result are same. But the test is still failing as if they were different.
In case 2, just the order of the elements in the collection is different. I am not sure why the test is failing because of that.
Here is my implementation and test case: https://github.com/TheLoneKing/eclipse-collections/commit/8f0025f244ee0284efe4a98f8ea6a811e90485c1
Can someone help on how to fix the failing tests?

That output doesn’t mean that the expected and actual were the same, just
that their toString() representations are the same. I’m guessing one is a
map of lists and one is a map of sets.

On Mon, Jul 22, 2019 at 9:38 PM Dineshkumar Gnanaprakasam <
notifications@github.com> wrote:

> I am trying to implement groupByAndCollect method on RichIterable. My
> test is failing with errors that looks confusing.
> Here are some errors:
>
>    1. [ERROR] ImmutableSortedBagImplTest.RichIterable_groupByAndCollect
>    expected: org.eclipse.collections.impl.map.mutable.UnifiedMap<{false=[12,
>    10, 8, 6, 4], true=[11, 9, 7, 5, 3]}> but was: org.eclipse.collections.impl.map.mutable.UnifiedMap<{false=[12,
>    10, 8, 6, 4], true=[11, 9, 7, 5, 3]}>
>    2. [ERROR] ImmutableHashBiMapInverseTest.RichIterable_groupByAndCollect
>    expected:<{false=[8, 10, 12, 4, 6], true=[9, 11, 3, 5, 7]}> but was:<{false=[4,
>    6, 8, 10, 12], true=[3, 5, 7, 9, 11]}>
>
> In case 1, both expected result and the actual result are same. But the
> test is still failing as if they were different.
> In case 2, just the order of the elements in the collection is different.
> I am not sure why the test is failing because of that.
> Here is my implementation and test case: TheLoneKing@8f0025f
> <https://github.com/TheLoneKing/eclipse-collections/commit/8f0025f244ee0284efe4a98f8ea6a811e90485c1>
> Can someone help on how to fix the failing tests?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse/eclipse-collections/issues/47?email_source=notifications&email_token=AAB3UIRSGQK6PS5FOQ6VQZLQAZOKXA5CNFSM4B22VGCKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2RUZTY#issuecomment-514018511>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAB3UISJW5UAG3BY7QG5W7DQAZOKXANCNFSM4B22VGCA>
> .
>

--- a/eclipse-collections-api/src/main/java/org/eclipse/collections/api/RichIterable.java
+++ b/eclipse-collections-api/src/main/java/org/eclipse/collections/api/RichIterable.java
@@ -2292,4 +2292,27 @@ public interface RichIterable<T>
      * @since 3.0
      */
     <K, V> MapIterable<K, V> aggregateBy(Function<? super T, ? extends K> groupBy, Function0<? extends V> zeroValueFactory, Function2<? super V, ? super T, ? extends V> nonMutatingAggregator);
+
+    /**
+     * Applies a groupBy function over the iterable, followed by a collect function.
+     *
+     * @param groupByFunction a {@link Function} to use as the groupBy transformation function
+     * @param collectFunction a {@link Function} to use as the collect transformation function
+     *
+     * @return The {@code target} collection where the key is the transformed result from applying the groupBy function
+     * and the value is the transformed result from applying the collect function.
+     *
+     * @see #groupBy(Function)
+     * @see Multimap#collectValues(Function)
+     *
+     * @since 10.1.0
+     */
+    default <K, V, R extends MutableMultimap<K, V>> R groupByAndCollect(
+            Function<? super T, ? extends K> groupByFunction,
+            Function<? super T, ? extends V> collectFunction,
+            R target)
+    {
+        this.forEach(each -> target.put(groupByFunction.apply(each), collectFunction.apply(each)));
+        return target;
+    }
 }

--- a/unit-tests-java8/src/test/java/org/eclipse/collections/test/RichIterableTestCase.java
+++ b/unit-tests-java8/src/test/java/org/eclipse/collections/test/RichIterableTestCase.java
@@ -56,6 +56,7 @@ import org.eclipse.collections.api.map.MapIterable;
 import org.eclipse.collections.api.map.MutableMap;
 import org.eclipse.collections.api.multimap.Multimap;
 import org.eclipse.collections.api.multimap.MutableMultimap;
+import org.eclipse.collections.api.multimap.list.MutableListMultimap;
 import org.eclipse.collections.api.partition.PartitionIterable;
 import org.eclipse.collections.api.tuple.Pair;
 import org.eclipse.collections.impl.Counter;
@@ -79,11 +80,13 @@ import org.eclipse.collections.impl.block.function.AddFunction;
 import org.eclipse.collections.impl.collector.Collectors2;
 import org.eclipse.collections.impl.factory.Bags;
 import org.eclipse.collections.impl.factory.Lists;
+import org.eclipse.collections.impl.factory.Multimaps;
 import org.eclipse.collections.impl.factory.Sets;
 import org.eclipse.collections.impl.factory.SortedSets;
 import org.eclipse.collections.impl.factory.primitive.ObjectDoubleMaps;
 import org.eclipse.collections.impl.factory.primitive.ObjectLongMaps;
 import org.eclipse.collections.impl.list.Interval;
+import org.eclipse.collections.impl.list.mutable.FastList;
 import org.eclipse.collections.impl.list.primitive.IntInterval;
 import org.eclipse.collections.impl.map.mutable.UnifiedMap;
 import org.eclipse.collections.impl.map.sorted.mutable.TreeSortedMap;
@@ -2085,6 +2088,22 @@ public interface RichIterableTestCase extends IterableTestCase
     {
         Object[] array = this.newWith(3, 3, 3, 2, 2, 1).toArray();
         assertEquals(Bags.immutable.with(3, 3, 3, 2, 2, 1), HashBag.newBagWith(array));
+    }
+
+    @Test
+    default void RichIterable_groupByAndCollect()
+    {
+        RichIterable<Integer> iterable = this.newWith(10, 9, 8, 7, 6, 5, 4, 3, 2, 1);
+        Function<Integer, Boolean> groupByFunction = integer -> IntegerPredicates.isOdd().accept(integer);
+        Function<Integer, Integer> collectFunction = integer -> integer + 2;
+
+        FastList<Integer> expectedOddNumberList = FastList.newListWith(3, 5, 7, 9, 11);
+        FastList<Integer> expectedEvenNumberList = FastList.newListWith(4, 6, 8, 10, 12);
+
+        MutableListMultimap<Boolean, Integer> targetResult = iterable.groupByAndCollect(groupByFunction, collectFunction, Multimaps.mutable.list.empty());
+
+        assertTrue(expectedOddNumberList.containsAll(targetResult.get(Boolean.TRUE)));
+        assertTrue(expectedEvenNumberList.containsAll(targetResult.get(Boolean.FALSE)));
     }
 
     class Holder<T extends Comparable<? super T>> implements Comparable<Holder<T>>

